docs/source/checkpointing.rst

```diff
@@ -8,29 +8,214 @@
 
 In addition to the ability to :doc:`write field data to vtu files
 <visualisation>`, suitable for visualisation in Paraview_, Firedrake
-has some support for checkpointing state to disk.  This enables
+has support for checkpointing state to disk.  This enables
 pausing, and subsequently resuming, a simulation at a later time.
 
-Restrictions on checkpointing
-=============================
+Checkpointing with CheckpointFile
+=================================
 
-The current support for checkpointing is somewhat limited.  One may
+:class:`~.CheckpointFile` class facilitates saving/loading meshes and
+:class:`~.Function` s to/from an HDF5_ file. 
+The implementation is scalable in that :class:`~.Function` s are
+saved to and loaded from the file entirely in parallel without needing
+to pass through a single process.
+It also supports flexible checkpointing, where one can save meshes and
+:class:`~.Function` s on :math:`N` processes and later load them on
+:math:`P` processes.
+
+Saving
+------
+
+In the following example we save in "example.h5" file two :class:`~.Function` s,
+along with the mesh on which they are defined.
+
+.. code-block:: python3
+
+    mesh = UnitSquareMesh(10, 10, name="meshA")
+    V = FunctionSpace(mesh, "CG", 2)
+    W = FunctionSpace(mesh, "CG", 1)
+    Z = V * W
+    f = Function(V, name="f")
+    g = Function(Z, name="g")
+    with CheckpointFile("example.h5", 'w') as afile:
+        afile.save_mesh(mesh)  # optional
+        afile.save_function(f)
+        afile.save_function(g)
+
+If the mesh name is not provided by the user when constructing the mesh, the
+default mesh name, :data:`~.DEFAULT_MESH_NAME`, is assigned, which is then
+used when saving in the file. We, however, strongly encourage users to name
+each mesh.
+
+Inspecting saved data
+---------------------
+
+Now "example.h5" file has been created and the mesh and :class:`~.Function`
+data have been saved.
+One can view the contents of the HDF5 file with "h5dump" utility shipped with
+the HDF5 installation; "h5dump -n example.h5", for instance, shows:
+
+::
+
+    HDF5 "example.h5" {
+    FILE_CONTENTS {
+     group      /
+     group      /topologies
+     group      /topologies/firedrake_mixed_meshes
+     group      /topologies/firedrake_mixed_meshes/meshA
+     group      /topologies/firedrake_mixed_meshes/meshA/firedrake_mixed_function_spaces
+     group      /topologies/firedrake_mixed_meshes/meshA/firedrake_mixed_function_spaces/firedrake_function_space_meshA_<CG2 on a triangle>_meshA_<CG1 on a triangle>
```

Comment:
Let's not have spaces and angle brackets.

Reply:
Yes! I will fix.